### PR TITLE
[stdlib] prevent MutableCollections from inappropriately inheriting a Slice<Self> subscript

### PR DESCRIPTION
**Pull request by glessard**
- adds a conditional protocol extension that defines `MutableCollection.MutableCollection.subscript(bounds: Range<_>) -> Slice<Self>` only when `SubSequence == Slice<Self>`.
- attempts to mark the unconditional extension method `MutableCollection.subscript(bounds: Range<_>) -> Slice<Self>` as unavailable
- `MutableCollection`s where `SubSequence == Slice<Self>` still get the expected default implementation.

- adds a default implementation of `MutableCollection.subscript(bounds: Range<_>) -> SubSequence`, marked unavailable, in order to prevent invalid conformances from compiling when `SubSequence` is not `Slice<Self>`.

Resolves SR-14850 / rdar://79898408 (and completes the fix to SR-14848).

### CODE REVIEW

**Comment by glessard** [comment]
Is there a wizard of availability annotations to help figure out what's needed in the last of these commits?

File: stdlib/public/core/MutableCollection.swift [hunk 1]
```diff
@@ -269,6 +284,45 @@ extension MutableCollection {
...
+  ///     let index = streetsSlice.firstIndex(of: "Evarts")    // 4
+  ///     streets[index!] = "Eustace"
+  ///     print(streets[index!])
+  ///     // Prints "Eustace"
+  ///
+  /// - Parameter bounds: A range of the collection's indices. The bounds of
+  ///   the range must be valid indices of the collection.
+  ///
+  /// - Complexity: O(1)
+  @inlinable
```
Comment (lorentey) [inline stdlib/public/core/MutableCollection.swift:314]: I think we'll want to mark this as `@_alwaysEmitIntoClient` and remove the availability annotation, to make it backdeployable. Unfortunately this seems unrelated to the issue at hand.

File: stdlib/public/core/MutableCollection.swift [hunk 2]
```diff
@@ -269,6 +284,45 @@ extension MutableCollection {
   }
 }
 
+@available(SwiftStdlib 5.5, *)
```
Comment (lorentey) [inline stdlib/public/core/MutableCollection.swift:287]: IIUC, we would only land this on `main`, so if we needed to add APIs with availability, we wouldn't be able to use the 5.5 cohort here. (We need to add a placeholder marker for the "next" version, whatever its number may become. `SwiftStdlib 9999` seems appropriate.)

File: stdlib/public/core/MutableCollection.swift [hunk 3]
```diff
@@ -238,6 +238,8 @@ extension MutableCollection {
   ///   the range must be valid indices of the collection.
   ///
   /// - Complexity: O(1)
+  // FIXME: somehow mark unavailable/deprecated/obsoleted for SwiftStdlib < 5.5
+  // @available(swift, unavailable)
```
Comment (lorentey) [inline stdlib/public/core/MutableCollection.swift:242]: I would expect that making this unconditionally unavailable (but still present) while adding the conditional extension below (with `_aeic`) would make things work.

File: stdlib/public/core/MutableCollection.swift (hunk 2 above)
Comment (glessard) [reply stdlib/public/core/MutableCollection.swift:287]: I hadn't tried `SwiftStdlib 9999`, thank you. I tried the list of OSs at 9999, and that did *not* work.

File: stdlib/public/core/MutableCollection.swift (hunk 3 above)
Comment (glessard) [reply stdlib/public/core/MutableCollection.swift:242]: That does work, but I wasn't sure whether it would be valid. Since we won't try to make it into the 5.5 branch, I'm less concerned.

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
@slavapestov I'd like your opinion on availability and ABI impact of these changes.

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
@swift-ci please smoke test

File: stdlib/public/core/MutableCollection.swift [hunk 4]
```diff
@@ -269,6 +283,46 @@ extension MutableCollection {
   }
 }
 
+@available(SwiftStdlib 9999, *)
```
Comment (lorentey) [inline stdlib/public/core/MutableCollection.swift:286]: I don't think this will work yet -- unless I'm mistaken, only 5.5 is defined in the CMake configs and the necessary build settings don't currently apply to the core stdlib.

In any case, I think we want to make these declarations backdeployable, in which case we don't want to declare availability for them.

File: stdlib/public/core/MutableCollection.swift (hunk 2 above)
Comment (lorentey) [reply stdlib/public/core/MutableCollection.swift]: IIRC the compiler doesn't understand that 9999 is supposed to sometimes be available -- so it only works when there is an explicit runtime availability check. (Or within declarations that are also 9999-available.)

File: stdlib/public/core/MutableCollection.swift (hunk 4 above)
Comment (lorentey) [reply stdlib/public/core/MutableCollection.swift:286]: I'm a bit surprised referring to an unknown platform / version here doesn't emit an error diagnostic -- @xymus, is that by design? 

(If so, that may actually come useful in Swift System. 🤔)

Comment (glessard) [reply stdlib/public/core/MutableCollection.swift:286]: Yeah, that one was a mystery. I'll just take it out and see what happens!

**Comment by glessard** [comment]
@swift-ci please smoke test macOS platform

File: stdlib/public/core/MutableCollection.swift (hunk 4 above)
Comment (xymus) [reply stdlib/public/core/MutableCollection.swift]: I was puzzled too so I checked. The compiler raises an error on a one word unknown availability macro, but only a warning on an unknown macro with a version number. This is because in case the macro doesn't match anything the compiler reverts to non-macro errors. We could clarify this in the future and adjust the diagnostic level at the same time.

You could use it in system until we have proper support in SwiftPM but that would likely mean a whole lot of warnings.

Comment (glessard) [reply stdlib/public/core/MutableCollection.swift]: Thanks

**Comment by glessard** [comment]
As it is, this change is tripping up the api-digester test. I believe that's the not-breaking-things test, so an adjustment is clearly needed.

**Comment by lorentey** [comment]
The source/abi compatibility tests aren't perfect -- they sometimes report false positives. We also know that this is _technically_ a source breaking change, so I'd expect we'll need to add at least a couple new expected differences.

#### Source compat diff

```
--- /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/api-digester/Output/stability-stdlib-source.swift.tmp.tmp/stability-stdlib-source.swift.expected	2021-07-23 13:23:30.000000000 -0700
+++ /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/api-digester/Output/stability-stdlib-source.swift.tmp.tmp/changes.txt.tmp	2021-07-23 13:23:30.000000000 -0700
@@ -1,4 +1,7 @@
+Accessor MutableCollection.subscript(_:).Get() has generic signature change from <Self where Self : Swift.MutableCollection> to <Self where Self : Swift.MutableCollection, Self.SubSequence == Swift.Slice<Self>>
+Accessor MutableCollection.subscript(_:).Set() has generic signature change from <Self where Self : Swift.MutableCollection> to <Self where Self : Swift.MutableCollection, Self.SubSequence == Swift.Slice<Self>>
+Subscript MutableCollection.subscript(_:) has generic signature change from <Self where Self : Swift.MutableCollection> to <Self where Self : Swift.MutableCollection, Self.SubSequence == Swift.Slice<Self>>
```

These are good -- the test correctly caught that we added an extra constraint on the `Slice`-returning subscript.

#### ABI diffs

```
--- /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/api-digester/Output/stability-stdlib-abi-with-asserts.test.tmp.tmp/stability-stdlib-abi.swift.expected.sorted	2021-07-23 13:23:31.000000000 -0700
+++ /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/test-macosx-x86_64/api-digester/Output/stability-stdlib-abi-with-asserts.test.tmp.tmp/changes.txt.tmp	2021-07-23 13:23:31.000000000 -0700
@@ -1,3 +1,4 @@
+Accessor _SmallString.subscript(_:).Modify() is a new API without @available attribute
```

This one is interesting -- it looks like the compiler generated an implicit `_modify` accessor for the smol string subscript. I did not know the compiler did that!

If true, then this is indeed an ABI break that needs to be fixed. A quick patch is to add an explicit `_modify` with an `@_alwaysEmitIntoClient` attribute.

**Comment by glessard** [comment]
@swift-ci please smoke test macOS platform

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
@swift-ci please smoke test macOS platform

**Comment by glessard** [comment]
@swift-ci please smoke test Linux platform

**Comment by glessard** [comment]
@lorentey Thanks for confirming I wasn't breaking anything I shouldn't break!
